### PR TITLE
Start working on TensorFlow 2.0 compatibility

### PR DESCRIPTION
**Pull request by gpengzhi**
This PR mainly contains `hyperparams.py`, `module_base.py`, `utils/mode.py`, and `utils/utils.py`.

`embedders` module added.
`networks` module added.
`encoders` module added.
`pretrained` module added.
`data` module added.
`tokenizer` module added.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ### New features
 
 * Support only Python 3.6 and 3.7. Drop support of older Python versions.
+* Support TensorFlow 2.0.
```
Comment (ZhitingHu) [inline CHANGELOG.md]: This is an old CHANGELOG.md

can we incorporate the new ?

File: README.md [hunk 2]
```diff
@@ -18,13 +18,14 @@
 * **Two Versions, (Mostly) Same Interfaces**. Texar-TensorFlow (this repo) and **[Texar-PyTorch](https://github.com/asyml/texar-pytorch)** have mostly the same interfaces. Both further combine the best design of TF and PyTorch:
```
Comment (ZhitingHu) [inline README.md]: Let's add "Texar TF2.0 is currently under incubation, please check back soon." to the beginning of README

**Comment by ZhitingHu** [comment]
Where is this file `docs/_static/img/texar_modules_big.png` from?

**Comment by gpengzhi** [comment]
> Where is this file `docs/_static/img/texar_modules_big.png` from?

Same picture we used in `texar-pytorch`.

File: setup.py [hunk 3]
```diff
@@ -1,21 +1,23 @@
...
 interfaces.
 '''
 
+if sys.version_info < (3, 6):
+    sys.exit('Python>=3.6 is required by Texar.')
+
 setuptools.setup(
     name="texar",
-    version="0.2.4-unreleased",
+    version="0.2.5-unreleased",
```
Comment (ZhitingHu) [inline setup.py]: make it `0.4-unreleased` so that `0.3` can be reserved for TF-1.0  -- once GPT2 and decoders of TF-1.0 are refactored we'd release Texar TF 0.3

I'm also thinking of make tf2.0 as `import texar.tf2 as tx`

File: texar/tf/hyperparams.py [hunk 4]
```diff
@@ -271,7 +267,7 @@ def _parse(hparams,
                 parsed_hparams[name] = value
             elif isinstance(value, type(default_value)):
                 parsed_hparams[name] = value
-            elif is_callable(value) and is_callable(default_value):
+            elif callable(value) and callable(default_value):
```
Comment (ZhitingHu) [inline texar/tf/hyperparams.py]: Can `callable` truly replace `is_callable`, which is a wrapper over `callable`? (https://github.com/asyml/texar/blob/master/texar/tf/utils/dtypes.py#L82)

File: texar/tf/module_base.py [hunk 5]
```diff
@@ -80,20 +50,20 @@ def __init__(self, hparams=None):
...
                     "`self._hparams` already exists. Argument `hparams` "
                     "must be set to `None` in this case.")
-        self._template = tf.make_template(self._hparams.name, self._build,
-                                          create_scope_now_=True)
-        self._unique_name = self.variable_scope.name.split("/")[-1]
-        self._trainable_variables = []
-        self._built = False
+
+    @abstractmethod
+    def call(self, inputs, *args, **kwargs):
```
Comment (ZhitingHu) [inline texar/tf/module_base.py]: How is the previous issue about `input` addressed now?

Comment (ZhitingHu) [reply texar/tf/module_base.py]: And how do other tools do?

File: texar/tf/version.py [hunk 6]
```diff
@@ -14,7 +14,7 @@
 
 _MAJOR = "0"
 _MINOR = "2"
-_REVISION = "4-unreleased"
+_REVISION = "5-unreleased"
```
Comment (ZhitingHu) [inline texar/tf/version.py]: 0.4.0-unreleased

File: texar/tf/hyperparams.py (hunk 4 above)
Comment (gpengzhi) [reply texar/tf/hyperparams.py]: I think it is true in Python 3.

File: texar/tf/module_base.py (hunk 5 above)
Comment (gpengzhi) [reply texar/tf/module_base.py]: To the best of my knowledge, we have two choices for our base class: `tf.Module` and `tf.keras.layers.Layer`. As pointed out in this [TensorFlow PR](https://github.com/tensorflow/community/pull/56), "Conceptually I see Keras `Layer` as enriched Modules with additional features (e.g. `build(shapes)`/`call(inputs)` pattern, json/yaml serialization, regularizers, losses, metrics, masking etc). If these features are useful to you, or if you are using your subclass in a Keras `Model` then Keras `Layer` is a good choice.

If the additional functionality of Keras is not useful for you then you may prefer to subclass `tf.Module` which is simpler with far fewer features. If `tf.Module` is not useful you may want to subclass `Checkpointable` or just `object`".

Comment (gpengzhi) [reply texar/tf/module_base.py]: https://github.com/tensorflow/community/blob/master/rfcs/20190117-tf-module.md

Comment (ZhitingHu) [reply texar/tf/module_base.py]: How is `tf.keras.layers.Layer` related to Keras `Model`? In particular, is `__call__(input, ..)` (i.e., name the 1st argument as `input`) a necessary interface in order to be compatible with Keras `Model`? If not, we may inherit `tf.keras.layers.Layer` but change this `__call__(input, ..)` interface a little bit

Comment (gpengzhi) [reply texar/tf/module_base.py]: `tf.keras.Model` is actually a subclass of `tf.keras.layers.Layer`, I think they have the same/similar interfaces.

File: texar/tf/modules/embedders/embedders.py [hunk 7]
```diff
@@ -51,7 +51,7 @@ class WordEmbedder(EmbedderBase):
             :meth:`default_hparams` for the hyperparameter structure and
             default values.
 
-    See :meth:`_build` for the inputs and outputs of the embedder.
+    See :meth:`call` for the inputs and outputs of the embedder.
```
Comment (ZhitingHu) [inline texar/tf/modules/embedders/embedders.py]: since we override `__call__` to have customized arguments, while the 1st argument of `call` is still `input` -- we shall expose `__call__` to users, instead of `call`. The readthedocs should include `__call__` and exclude `call`

File: texar/tf/modules/embedders/embedders.py [hunk 8]
```diff
@@ -61,8 +61,9 @@ class WordEmbedder(EmbedderBase):
             soft_ids = tf.random_uniform(shape=[32, 10, 100])
 
             embedder = WordEmbedder(vocab_size=100, hparams={'dim': 256})
-            ids_emb = embedder(ids=ids) # shape: [32, 10, 256]
-            soft_ids_emb = embedder(soft_ids=soft_ids) # shape: [32, 10, 256]
+            ids_emb = embedder(ids) # shape: [32, 10, 256]
```
Comment (ZhitingHu) [inline texar/tf/modules/embedders/embedders.py]: Given that we have over-writen `__call__`, this should not be changed

File: texar/tf/modules/embedders/embedders.py [hunk 9]
```diff
@@ -215,11 +209,16 @@ def _build(self, ids=None, soft_ids=None, mode=None, **kwargs):
             and ``shape(embedding) = [vocab_size, emb_dim]``, then the return
             tensor has shape ``[batch_size, max_time, emb_dim]``.
         """
-        if ids is not None:
+        return super().__call__(ids, soft_ids, mode, **kwargs)
```
Comment (ZhitingHu) [inline texar/tf/modules/embedders/embedders.py]: Is this correct? Either `ids` or `soft_ids` is `inputs`, right?

File: texar/tf/modules/embedders/position_embedders.py [hunk 10]
```diff
@@ -147,28 +138,30 @@ def _build(self, positions=None, sequence_length=None, mode=None, **kwargs):
         Returns:
             A `Tensor` of shape `shape(inputs) + embedding dimension`.
         """
+        return super().__call__(positions, sequence_length, mode, **kwargs)
```
Comment (ZhitingHu) [inline texar/tf/modules/embedders/position_embedders.py]: Same here, either `positions` or `sequence_length` is `inputs` ?

**Comment by gpengzhi** [comment]
Close this PR for now for development, Will reopen it soon.
